File: webapp/lib/gemini.ts
import "server-only";

const GEMINI_URL =
  "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={key}";

const DEFAULT_MODEL = "gemini-2.5-flash";

export function geminiAvailable(): boolean {
  return !!process.env.GEMINI_API_KEY;
}

type GenOpts = {
  system: string;
  user: string;
  json?: boolean;
  temperature?: number;
  maxTokens?: number;
  model?: string;
  useSearch?: boolean;
};

export async function geminiGen(opts: GenOpts): Promise<string> {
  const key = process.env.GEMINI_API_KEY;
  if (!key) throw new Error("GEMINI_API_KEY is not set");

  const body: Record<string, unknown> = {
    systemInstruction: { role: "system", parts: [{ text: opts.system }] },
    contents: [{ role: "user", parts: [{ text: opts.user }] }],
    generationConfig: {
      temperature: opts.temperature ?? 0.7,
      maxOutputTokens: opts.maxTokens ?? 1024,
      ...(opts.json ? { responseMimeType: "application/json" } : {}),
    },
    ...(opts.useSearch ? { tools: [{ google_search: {} }] } : {}),
  };

  const url = GEMINI_URL.replace("{model}", opts.model ?? DEFAULT_MODEL).replace(
    "{key}",
    key,
  );
  const r = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
    // Vercel functions have 30s default; leave room
    signal: AbortSignal.timeout(45_000),
  });
  if (!r.ok) {
    const text = await r.text().catch(() => "");
    throw new Error(`gemini ${r.status}: ${text.slice(0, 300)}`);
  }
  const data = (await r.json()) as {
    candidates?: { content?: { parts?: { text?: string }[] } }[];
  };
  const parts = data.candidates?.[0]?.content?.parts ?? [];
  return parts.map((p) => p.text ?? "").join("");
}

export async function geminiJSON<T>(opts: GenOpts): Promise<T> {
  const raw = await geminiGen({ ...opts, json: true });
  return JSON.parse(raw) as T;
}

File: bot/src/services/gemini.py
"""
Gemini client (Google AI Studio API).
Used as the primary AI for invites / place suggestions when GEMINI_API_KEY is set.
"""
from __future__ import annotations

import json
from typing import Any

import httpx

from .. import config

GEMINI_URL = (
    "https://generativelanguage.googleapis.com/v1beta/models/"
    "{model}:generateContent?key={key}"
)
DEFAULT_MODEL = "gemini-2.5-flash"


def available() -> bool:
    return bool(config.GEMINI_API_KEY)


async def complete(
    system: str,
    user: str,
    *,
    json_mode: bool = False,
    temperature: float = 0.8,
    max_tokens: int = 1024,
    model: str = DEFAULT_MODEL,
    use_search: bool = False,
) -> str:
    if not config.GEMINI_API_KEY:
        raise RuntimeError("GEMINI_API_KEY is not set")

    payload: dict[str, Any] = {
        "systemInstruction": {"role": "system", "parts": [{"text": system}]},
        "contents": [{"role": "user", "parts": [{"text": user}]}],
        "generationConfig": {
            "temperature": temperature,
            "maxOutputTokens": max_tokens,
        },
    }
    if json_mode:
        payload["generationConfig"]["responseMimeType"] = "application/json"
    if use_search:
        payload["tools"] = [{"google_search": {}}]

    url = GEMINI_URL.format(model=model, key=config.GEMINI_API_KEY)
    async with httpx.AsyncClient(timeout=45) as client:
        r = await client.post(url, json=payload)
        r.raise_for_status()
        data = r.json()

    candidates = data.get("candidates") or []
    if not candidates:
        raise RuntimeError(f"Gemini returned no candidates: {data}")
    parts = candidates[0].get("content", {}).get("parts") or []
    return "".join(p.get("text", "") for p in parts)


async def json_complete(system: str, user: str, **kw: Any) -> dict[str, Any]:
    raw = await complete(system, user, json_mode=True, **kw)
    return json.loads(raw)
